lib/src/rules/do_not_create_setter_without_getter.dart

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.avoid_setter_without_getter;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/ast.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+const _desc = r' ';
+
+const _details = r'''
+
+**DON’T** define a setter without a corresponding getter.
+
+**BAD:**
+```
+class Bad {
+  int l, r;
+
+  set length(int newLength) {
+    r = l + length;
+  }
+}
+```
+
+**GOOD:**
+```
+class Good {
+  int l, r;
+
+  int get length => r - l;
+
+  set length(int newLength) {
+    r = l + length;
+  }
+}
+```
+
+''';
+
+bool _hasInheritedGetter(MethodDeclaration node) =>
+    DartTypeUtilities.lookUpInheritedConcreteGetter(node) != null;
+
+bool _hasInheritedSetter(MethodDeclaration node) =>
+    DartTypeUtilities.lookUpInheritedConcreteSetter(node) != null;
+
+class AvoidSetterWithoutGetter extends LintRule {
+  _Visitor _visitor;
+  AvoidSetterWithoutGetter()
+      : super(
+            name: 'do_not_create_setter_without_getter',
+            description: _desc,
+            details: _details,
+            group: Group.style) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  visitClassDeclaration(ClassDeclaration node) {
+    final getters = new Set<String>();
+    final setters = <String, MethodDeclaration>{};
+
+    // Build getter/setter maps
+    final methods = node.members.where(isMethod);
+    for (MethodDeclaration method in methods) {
+      if (method.isGetter) {
+        getters.add(method.name.toString());
+      } else if (method.isSetter &&
+          !_hasInheritedSetter(method) &&
+          !_hasInheritedGetter(method)) {
+        setters[method.name.toString()] = method;
+      }
+    }
+    void _reportLintFromKey(String id) {
+      rule.reportLint(setters[id]);
+    }
+
+    setters.keys.toSet().difference(getters).forEach(_reportLintFromKey);
```

Comment:
I believe that there is API for looking up the declaration of a getter in either a class or its superclasses, in which case there's no reason to build up collections. You could just iterate over the setters and ensure that either there is an overridden setter or that there is a getter, and if not report the lint immediately.